File: src/chat/JavaspectreCommandBlock.cpp
// src/chat/JavaspectreCommandBlock.cpp
// Javaspectre Command Block for C++ Chat Integration
// © 2025 Perplexity Labs Inc. / Dr. Jacob S. Farmer

#include <string>
#include <string_view>
#include <unordered_map>
#include <functional>
#include <vector>
#include <chrono>
#include <random>
#include <sstream>
#include <iomanip>
#include <optional>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

// ---------- Core types ----------

struct ChatCommandContext {
    std::string input;
    std::vector<std::string> args;
    std::string userId;
    std::string sessionId;
    json metadata;
};

using CommandHandler = std::function<json(const ChatCommandContext&)>;

class ChatInterface {
public:
    virtual ~ChatInterface() = default;
    virtual void registerCommand(const std::string& trigger, CommandHandler handler) = 0;
    virtual void log(const std::string& message, const json& meta = json::object()) = 0;
};

// ---------- Utility functions ----------

static std::string nowIso() {
    using namespace std::chrono;
    const auto now = system_clock::now();
    const auto tt  = system_clock::to_time_t(now);
    std::tm tm{};
#if defined(_WIN32)
    gmtime_s(&tm, &tt);
#else
    gmtime_r(&tt, &tm);
#endif
    std::ostringstream oss;
    oss << std::put_time(&tm, "%Y-%m-%dT%H:%M:%SZ");
    return oss.str();
}

static std::string createTraceId() {
    using namespace std::chrono;
    const auto now = system_clock::now().time_since_epoch();
    const auto ms  = duration_cast<milliseconds>(now).count();
    std::mt19937_64 rng(static_cast<std::mt19937_64::result_type>(ms));
    std::uniform_int_distribution<uint64_t> dist;
    const uint64_t r = dist(rng);

    std::ostringstream oss;
    oss << "jsp-" << std::hex << ms << "-" << (r & 0xFFFFFFull);
    return oss.str();
}

static void safeLog(ChatInterface* chat, const std::string& message, const json& meta = json::object()) {
    if (chat) {
        chat->log(message, meta);
    }
}

// ---------- Risk classification ----------

struct RiskClassification {
    bool cognitiveHazard{false};
    bool entropyAnomaly{false};
    bool ontologicalInstability{false};
    std::vector<json> signals;
};

class ALNRiskClassifier {
public:
    RiskClassification classify(std::string_view input, const json& meta) const {
        const std::string textLower = toLower(std::string(input));

        const std::vector<std::string> cognitiveHazardKeywords{
            "forbidden", "memetic", "cursed", "anomalous cognition"
        };
        const std::vector<std::string> entropyKeywords{
            "random stream", "noise", "entropy source", "unstable log"
        };
        const std::vector<std::string> ontologicalKeywords{
            "reality rewrite", "self-erasure", "identity collapse", "ontology loop"
        };

        RiskClassification rc;

        if (containsAny(textLower, cognitiveHazardKeywords)) {
            rc.cognitiveHazard = true;
            rc.signals.push_back(json{
                {"type", "cognitive-hazard-indicator"},
                {"reason", "Detected memetic/forbidden semantics in input text."},
                {"weight", 0.9}
            });
        }

        if (containsAny(textLower, entropyKeywords)) {
            rc.entropyAnomaly = true;
            rc.signals.push_back(json{
                {"type", "entropy-anomaly-indicator"},
                {"reason", "Detected references to entropy/noise sources."},
                {"weight", 0.7}
            });
        }

        if (containsAny(textLower, ontologicalKeywords)) {
            rc.ontologicalInstability = true;
            rc.signals.push_back(json{
                {"type", "ontological-instability-indicator"},
                {"reason", "Detected ontology/identity destabilizing language."},
                {"weight", 0.85}
            });
        }

        if (meta.contains("layer") && meta["layer"] == "deep-excavation") {
            rc.signals.push_back(json{
                {"type", "deep-excavation-context"},
                {"reason", "Context flagged as deep-excavation; raising review priority."},
                {"weight", 0.4}
            });
        }

        return rc;
    }

private:
    static std::string toLower(std::string s) {
        for (auto& ch : s) {
            ch = static_cast<char>(::tolower(static_cast<unsigned char>(ch)));
        }
        return s;
    }

    static bool containsAny(const std::string& text, const std::vector<std::string>& words) {
        for (const auto& w : words) {
            if (text.find(w) != std::string::npos) {
                return true;
            }
        }
        return false;
    }
};

// ---------- Synthesis result ----------

struct SynthesisResult {
    std::string blueprint;
    std::string integrationPotential; // "Low" | "Medium" | "High" | "Unknown"
    bool complianceAnchor{true};
};

// ---------- JavaspectreCommandBlock ----------

class JavaspectreCommandBlock {
public:
    explicit JavaspectreCommandBlock(ChatInterface* chat = nullptr)
        : chatInterface_(chat), version_("1.0.0") {
        if (chatInterface_) {
            init(chatInterface_);
        }
    }

    void init(ChatInterface* chat) {
        chatInterface_ = chat;
        safeLog(chatInterface_, "Command Block Initialized.", {
            {"module", name_},
            {"version", version_}
        });
        attachCommands();
    }

private:
    std::string name_{"JavaspectreCommandBlock"};
    std::string version_;
    bool active_{true};
    ChatInterface* chatInterface_{nullptr};
    ALNRiskClassifier riskClassifier_{};

    void attachCommands() {
        if (!chatInterface_) {
            throw std::runtime_error("[Javaspectre] Cannot attach commands before initialization.");
        }

        registerCommand("/excavate", [this](const ChatCommandContext& ctx) {
            return runExcavation(ctx);
        });

        registerCommand("/spectral-scan", [this](const ChatCommandContext& ctx) {
            return runSpectralScan(ctx);
        });

        registerCommand("/classify", [this](const ChatCommandContext& ctx) {
            return runClassification(ctx);
        });

        registerCommand("/synthesize", [this](const ChatCommandContext& ctx) {
            return runSynthesis(ctx);
        });

        safeLog(chatInterface_, "Commands registered.", {
            {"commands", json::array({"/excavate", "/spectral-scan", "/classify", "/synthesize"})}
        });
    }

    void registerCommand(const std::string& trigger, CommandHandler handler) {
        chatInterface_->registerCommand(trigger, std::move(handler));
    }

    // ----- Standardized envelope -----

    json outputEnvelope(
        const json& data,
        const std::string& title,
        const std::string& command,
        const ChatCommandContext& ctx,
        const std::optional<std::string>& layer = std::nullopt
    ) const {
        json meta{
            {"version", version_},
            {"command", command},
            {"userId", ctx.userId},
            {"sessionId", ctx.sessionId},
            {"traceId", createTraceId()}
        };
        if (layer.has_value()) {
            meta["layer"] = *layer;
        }

        json envelope{
            {"title", title},
            {"timestamp", nowIso()},
            {"system", "Javaspectre Command Block (C++)"},
            {"data", data},
            {"meta", meta}
        };
        return envelope;
    }

    // ----- Command handlers -----

    json runExcavation(const ChatCommandContext& ctx) const {
        const std::string layer = ctx.args.empty() ? "default" : ctx.args.front();

        // Placeholder for actual excavation engine integration.
        json coreResult{
            {"layer", layer},
            {"status", "ok"},
            {"notes", "Integrate with C++ native VirtualObjectExcavator or bridge to JS/ALN layer."}
        };

        // Best-effort JSON parse of input to support deep virtual-object excavation.
        json parsed = json::object();
        if (!ctx.input.empty()) {
            try {
                parsed = json::parse(ctx.input);
            } catch (...) {
                parsed = json{
                    {"status", "skipped"},
                    {"reason", "Input is not valid JSON; deep excavation not performed."}
                };
            }
        }

        json data{
            {"layer", layer},
            {"coreExcavation", coreResult},
            {"virtualObjectInput", parsed}
        };

        return outputEnvelope(data, "Excavation Report", "/excavate", ctx, layer);
    }

    json runSpectralScan(const ChatCommandContext& ctx) const {
        const std::string trimmed = trim(ctx.input);

        if (trimmed.empty()) {
            json data{{"error", "No input provided for spectral scan."}};
            return outputEnvelope(data, "Spectral Scan Error", "/spectral-scan", ctx);
        }

        // Minimal semantic + structural hints; extend with full ALN scan pipeline.
        json scan{
            {"length", trimmed.size()},
            {"hasJsonBraces", trimmed.find('{') != std::string::npos},
            {"hasCodeLikeTokens", trimmed.find("class ") != std::string::npos ||
                                  trimmed.find("function ") != std::string::npos},
            {"preview", trimmed.substr(0, std::min<size_t>(80, trimmed.size()))}
        };

        json data{
            {"scan", scan},
            {"capabilityHints", json::array({"spectral-analysis", "structure-detection"})}
        };

        return outputEnvelope(data, "Spectral Scan Output", "/spectral-scan", ctx);
    }

    json runClassification(const ChatCommandContext& ctx) const {
        const std::string trimmed = trim(ctx.input);
        json meta = ctx.metadata;
        if (meta.is_null()) {
            meta = json::object();
        }
        if (!ctx.args.empty()) {
            meta["layer"] = ctx.args.front();
        }

        const auto rc = riskClassifier_.classify(trimmed, meta);

        json data{
            {"cognitiveHazard", rc.cognitiveHazard},
            {"entropyAnomaly", rc.entropyAnomaly},
            {"ontologicalInstability", rc.ontologicalInstability},
            {"signals", rc.signals}
        };

        return outputEnvelope(data, "Risk & Anomaly Classification", "/classify", ctx);
    }

    json runSynthesis(const ChatCommandContext& ctx) const {
        const std::string trimmed = trim(ctx.input);

        SynthesisResult sr;
        if (!trimmed.empty()) {
            sr.blueprint = "Blueprint constructed for: " + trimmed;
            sr.integrationPotential = "High";
            sr.complianceAnchor = true;
        } else {
            sr.blueprint = "Blueprint constructed for: <empty-input>";
            sr.integrationPotential = "Unknown";
            sr.complianceAnchor = true;
        }

        json data{
            {"blueprint", sr.blueprint},
            {"integrationPotential", sr.integrationPotential},
            {"complianceAnchor", sr.complianceAnchor}
        };

        return outputEnvelope(data, "Synthesis Protocol", "/synthesize", ctx);
    }

    // ----- Helpers -----

    static std::string trim(const std::string& s) {
        if (s.empty()) return s;
        std::size_t start = 0;
        while (start < s.size() && std::isspace(static_cast<unsigned char>(s[start]))) {
            ++start;
        }
        if (start == s.size()) return std::string();
        std::size_t end = s.size() - 1;
        while (end > start && std::isspace(static_cast<unsigned char>(s[end]))) {
            --end;
        }
        return s.substr(start, end - start + 1);
    }
};

File: src/chat/JavaspectreCommandBlock.js
// src/chat/JavaspectreCommandBlock.js
// Javaspectre Command Block for AI Chat Integration
// © 2025 Perplexity Labs Inc. / Dr. Jacob S. Farmer

import {
  SpectralScanner,
  ObjectExcavator,
} from "../spectral-core/spectral-tools.js";

import { VirtualObjectExcavator } from "../core/VirtualObjectExcavator.js";
import { JavaspectreCapabilities } from "../capabilities/JavaspectreCapabilities.js";

/**
 * Core types
 */

export class JavaspectreCommandBlockContext {
  constructor({ input, args = [], userId, sessionId, metadata = {} }) {
    this.input = typeof input === "string" ? input : "";
    this.args = Array.isArray(args) ? args : [];
    this.userId = userId;
    this.sessionId = sessionId;
    this.metadata = metadata;
  }
}

/**
 * Chat interface abstraction
 */
export class JavaspectreChatInterfaceAdapter {
  constructor(host) {
    this.host = host;
  }

  registerCommand(trigger, handler) {
    if (!this.host || typeof this.host.registerCommand !== "function") {
      throw new Error("[Javaspectre] Host chat interface does not support registerCommand.");
    }
    this.host.registerCommand(trigger, async (ctx) => {
      const wrapped = new JavaspectreCommandBlockContext(ctx);
      return handler(wrapped);
    });
  }

  log(message, meta = {}) {
    if (this.host && typeof this.host.log === "function") {
      this.host.log(message, meta);
      return;
    }
    try {
      console.log(`[Javaspectre] ${message}`, meta);
    } catch {
      // no-op
    }
  }
}

/**
 * Risk & synthesis types
 */

export class JavaspectreRiskClassification {
  constructor({ cognitiveHazard, entropyAnomaly, ontologicalInstability, signals }) {
    this.cognitiveHazard = Boolean(cognitiveHazard);
    this.entropyAnomaly = Boolean(entropyAnomaly);
    this.ontologicalInstability = Boolean(ontologicalInstability);
    this.signals = Array.isArray(signals) ? signals : [];
  }
}

export class JavaspectreSynthesisResult {
  constructor({ blueprint, integrationPotential, complianceAnchor }) {
    this.blueprint = blueprint;
    this.integrationPotential = integrationPotential;
    this.complianceAnchor = Boolean(complianceAnchor);
  }
}

/**
 * Output envelope
 */

export class JavaspectreOutputEnvelope {
  constructor({ title, command, data, userId, sessionId, layer, version }) {
    this.title = title;
    this.timestamp = new Date().toISOString();
    this.system = "Javaspectre Command Block";
    this.data = data;
    this.meta = {
      version,
      command,
      userId,
      sessionId,
      layer,
      traceId: JavaspectreOutputEnvelope.createTraceId(),
    };
  }

  static createTraceId() {
    return `jsp-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`;
  }
}

/**
 * Deterministic-ish risk classifier using ALN-style heuristics
 * (no randomness; pure function of input text + metadata).
 */
export class ALNRiskClassifier {
  classify(input, metadata = {}) {
    const text = (input || "").toLowerCase();
    const signals = [];

    const cognitiveHazardKeywords = ["forbidden", "memetic", "cursed", "anomalous cognition"];
    const entropyKeywords = ["random stream", "noise", "entropy source", "unstable log"];
    const ontologicalKeywords = ["reality rewrite", "self-erasure", "identity collapse", "ontology loop"];

    const containsAny = (words) => words.some((w) => text.includes(w));

    const cognitiveHazard = containsAny(cognitiveHazardKeywords);
    if (cognitiveHazard) {
      signals.push({
        type: "cognitive-hazard-indicator",
        reason: "Detected memetic/forbidden semantics in input text.",
        weight: 0.9,
      });
    }

    const entropyAnomaly = containsAny(entropyKeywords);
    if (entropyAnomaly) {
      signals.push({
        type: "entropy-anomaly-indicator",
        reason: "Detected references to entropy/noise sources.",
        weight: 0.7,
      });
    }

    const ontologicalInstability = containsAny(ontologicalKeywords);
    if (ontologicalInstability) {
      signals.push({
        type: "ontological-instability-indicator",
        reason: "Detected ontology/identity destabilizing language.",
        weight: 0.85,
      });
    }

    if (metadata && metadata.layer === "deep-excavation") {
      signals.push({
        type: "deep-excavation-context",
        reason: "Context flagged as deep-excavation; raising review priority.",
        weight: 0.4,
      });
    }

    return new JavaspectreRiskClassification({
      cognitiveHazard,
      entropyAnomaly,
      ontologicalInstability,
      signals,
    });
  }
}

/**
 * Deep virtual-object excavation bridge for chat environments.
 * - Accepts raw JSON / JS-like input.
 * - Feeds into VirtualObjectExcavator.
 * - Returns a compact, chat-friendly summary + raw model.
 */
export class ChatVirtualObjectBridge {
  constructor() {
    this.excavator = new VirtualObjectExcavator({
      maxDepth: 6,
      maxArraySample: 10,
      includeDom: false,
      includeFunctions: false,
    });
  }

  safeParse(input) {
    const trimmed = (input || "").trim();
    if (!trimmed) return null;
    try {
      // Best-effort JSON parse for chat payloads.
      return JSON.parse(trimmed);
    } catch {
      return null;
    }
  }

  excavateFromInput(input) {
    const value = this.safeParse(input);
    if (value === null) {
      return {
        summary: {
          status: "skipped",
          reason: "Input is not parseable as JSON; no deep excavation performed.",
        },
        virtualObjects: [],
        relationships: [],
        domSheets: [],
      };
    }

    const result = this.excavator.excavate({ value, domRoot: null });

    const dominantKinds = {};
    for (const vo of result.virtualObjects) {
      const kind = vo.kind || "unknown";
      dominantKinds[kind] = (dominantKinds[kind] || 0) + 1;
    }

    return {
      summary: {
        status: "ok",
        rootId: result.root && result.root.id,
        objectCount: result.virtualObjects.length,
        relationshipCount: result.relationships.length,
        domSheetCount: result.domSheets.length,
        dominantKinds,
      },
      virtualObjects: result.virtualObjects,
      relationships: result.relationships,
      domSheets: result.domSheets,
    };
  }
}

/**
 * Javaspectre Command Block
 */
export class JavaspectreCommandBlock {
  constructor(chatInterface) {
    this.name = "JavaspectreCommandBlock";
    this.version = "1.1.0";
    this.active = true;

    this.chatInterface = chatInterface
      ? new JavaspectreChatInterfaceAdapter(chatInterface)
      : undefined;

    this.riskClassifier = new ALNRiskClassifier();
    this.voBridge = new ChatVirtualObjectBridge();

    if (this.chatInterface) {
      this.init(this.chatInterface);
    }
  }

  init(chatInterface) {
    this.chatInterface = chatInterface instanceof JavaspectreChatInterfaceAdapter
      ? chatInterface
      : new JavaspectreChatInterfaceAdapter(chatInterface);

    this.chatInterface.log("Command Block Initialized.", {
      module: this.name,
      version: this.version,
    });

    this.attachCommands();
  }

  attachCommands() {
    if (!this.chatInterface) {
      throw new Error("[Javaspectre] Cannot attach commands before initialization.");
    }

    const commands = [
      { trigger: "/excavate", handler: this.runExcavation.bind(this) },
      { trigger: "/spectral-scan", handler: this.runSpectralScan.bind(this) },
      { trigger: "/classify", handler: this.runClassification.bind(this) },
      { trigger: "/synthesize", handler: this.runSynthesis.bind(this) },
      { trigger: "/capabilities", handler: this.runCapabilities.bind(this) },
    ];

    for (const cmd of commands) {
      this.chatInterface.registerCommand(cmd.trigger, cmd.handler);
    }

    this.chatInterface.log("Commands registered.", {
      commands: commands.map((c) => c.trigger),
    });
  }

  /**
   * /excavate – layer-based + optional JSON deep virtual-object excavation.
   * Usage examples inside chat:
   *   /excavate default { "user": { "id": 1, "roles": ["admin"] } }
   *   /excavate deep-excavation { ...large-structure... }
   */
  async runExcavation(context) {
    const layer = context.args && context.args[0]
      ? String(context.args[0]).trim()
      : "default";

    const payload = context.input || "";
    const deepModel = this.voBridge.excavateFromInput(payload);

    try {
      const coreResult = await Promise.resolve(ObjectExcavator.excavate(layer));

      return new JavaspectreOutputEnvelope({
        title: "Excavation Report",
        command: "/excavate",
        userId: context.userId,
        sessionId: context.sessionId,
        layer,
        version: this.version,
        data: {
          layer,
          coreExcavation: coreResult,
          virtualObjectExcavation: deepModel,
        },
      });
    } catch (error) {
      this.chatInterface && this.chatInterface.log("Excavation failed.", {
        error: String(error),
        layer,
      });

      return new JavaspectreOutputEnvelope({
        title: "Excavation Error",
        command: "/excavate",
        userId: context.userId,
        sessionId: context.sessionId,
        layer,
        version: this.version,
        data: {
          error: "Excavation failed",
          details: String(error),
        },
      });
    }
  }

  /**
   * /spectral-scan – semantic & structural scanning with capability hints.
   */
  async runSpectralScan(context) {
    const input = (context.input || "").trim();

    if (!input) {
      return new JavaspectreOutputEnvelope({
        title: "Spectral Scan Error",
        command: "/spectral-scan",
        userId: context.userId,
        sessionId: context.sessionId,
        version: this.version,
        data: { error: "No input provided for spectral scan." },
      });
    }

    try {
      const scan = await Promise.resolve(SpectralScanner.performFullScan(input));

      const matchedCapabilities = this.matchCapabilities(input);

      return new JavaspectreOutputEnvelope({
        title: "Spectral Scan Output",
        command: "/spectral-scan",
        userId: context.userId,
        sessionId: context.sessionId,
        version: this.version,
        data: {
          scan,
          matchedCapabilities,
        },
      });
    } catch (error) {
      this.chatInterface && this.chatInterface.log("Spectral scan failed.", {
        error: String(error),
      });

      return new JavaspectreOutputEnvelope({
        title: "Spectral Scan Error",
        command: "/spectral-scan",
        userId: context.userId,
        sessionId: context.sessionId,
        version: this.version,
        data: {
          error: "Spectral scan failed",
          details: String(error),
        },
      });
    }
  }

  /**
   * /classify – deterministic risk & anomaly classification using ALNRiskClassifier.
   */
  async runClassification(context) {
    const input = (context.input || "").trim();
    const analysis = this.riskClassifier.classify(input, {
      userId: context.userId,
      sessionId: context.sessionId,
      layer: context.args && context.args[0] ? String(context.args[0]) : undefined,
    });

    return new JavaspectreOutputEnvelope({
      title: "Risk & Anomaly Classification",
      command: "/classify",
      userId: context.userId,
      sessionId: context.sessionId,
      version: this.version,
      data: analysis,
    });
  }

  /**
   * /synthesize – blueprint + integration potential, preserving your original semantics.
   */
  async runSynthesis(context) {
    const input = (context.input || "").trim();

    const base = new JavaspectreSynthesisResult({
      blueprint: input
        ? `Blueprint constructed for: ${input}`
        : "Blueprint constructed for: <empty-input>",
      integrationPotential: input ? "High" : "Unknown",
      complianceAnchor: true,
    });

    return new JavaspectreOutputEnvelope({
      title: "Synthesis Protocol",
      command: "/synthesize",
      userId: context.userId,
      sessionId: context.sessionId,
      version: this.version,
      data: base,
    });
  }

  /**
   * /capabilities – expose the 15 spectral capabilities in a chat-friendly format.
   */
  async runCapabilities(context) {
    const capabilities = Array.isArray(JavaspectreCapabilities)
      ? JavaspectreCapabilities
      : [];

    return new JavaspectreOutputEnvelope({
      title: "Javaspectre Capabilities Map",
      command: "/capabilities",
      userId: context.userId,
      sessionId: context.sessionId,
      version: this.version,
      data: {
        count: capabilities.length,
        capabilities,
      },
    });
  }

  /**
   * Simple intent-to-capability matcher using text heuristics.
   */
  matchCapabilities(input) {
    const text = (input || "").toLowerCase();
    const capabilities = Array.isArray(JavaspectreCapabilities)
      ? JavaspectreCapabilities
      : [];

    const matches = [];

    for (const cap of capabilities) {
      const tags = Array.isArray(cap.spectralTags) ? cap.spectralTags : [];
      const keywords = [
        cap.name || "",
        cap.category || "",
        ...(tags || []),
      ]
        .join(" ")
        .toLowerCase()
        .split(/[\s,]+/)
        .filter(Boolean);

      const score = keywords.reduce((acc, kw) => {
        if (!kw) return acc;
        if (text.includes(kw)) return acc + 1;
        return acc;
      }, 0);

      if (score > 0) {
        matches.push({
          id: cap.id,
          name: cap.name,
          category: cap.category,
          heuristicScore: cap.heuristicScore || 0,
          matchScore: score,
          primaryCli: cap.primaryCli,
        });
      }
    }

    matches.sort((a, b) => {
      const h = (b.heuristicScore || 0) - (a.heuristicScore || 0);
      if (h !== 0) return h;
      return (b.matchScore || 0) - (a.matchScore || 0);
    });

    return matches.slice(0, 5);
  }
}

/**
 * Auto-mount for AI chat environments supporting custom modules.
 */
// eslint-disable-next-line no-undef
const globalAny = globalThis;

if (typeof globalAny.ChatAI !== "undefined") {
  const block = new JavaspectreCommandBlock(globalAny.ChatAI);
  // block.init is called from constructor when chatInterface is provided.
}

export default JavaspectreCommandBlock;
